packages/builders/src/messages/Assertions.ts

```diff
@@ -73,5 +82,47 @@ export const messagePredicate = z
 				(data.sticker_ids !== undefined && data.sticker_ids.length > 0)
 			);
 		},
-		{ message: 'Messages must have content, embeds, a poll, attachments, components, or stickers' },
+		{ message: 'Messages must have content, embeds, a poll, attachments, components OR stickers' },
 	);
+
+const allTopLevelComponentsPredicate = z
+	.union([
+		basicActionRowPredicate,
+		z.object({
+			type: z.union([
+				// Components v1
+				z.literal(ComponentType.Button),
+				z.literal(ComponentType.ChannelSelect),
+				z.literal(ComponentType.MentionableSelect),
+				z.literal(ComponentType.RoleSelect),
+				z.literal(ComponentType.StringSelect),
+				z.literal(ComponentType.UserSelect),
+
+				// Components v2
+				z.literal(ComponentType.Container),
+				z.literal(ComponentType.File),
+				z.literal(ComponentType.MediaGallery),
+				z.literal(ComponentType.Section),
+				z.literal(ComponentType.Separator),
+				z.literal(ComponentType.TextDisplay),
+				z.literal(ComponentType.Thumbnail),
+			]),
+		}),
+	])
+	.array()
+	.min(1)
+	.max(10);
+
+const messageComponentsV2Predicate = baseMessagePredicate.extend({
+	components: allTopLevelComponentsPredicate,
+	flags: z.number().refine((flags) => {
+		return (flags & MessageFlags.IsComponentsV2) === MessageFlags.IsComponentsV2;
+	}),
+	// These fields cannot be set
+	content: z.string().length(0).nullish(),
+	embeds: z.array(embedPredicate).max(0).nullish(),
```

Comment:
```suggestion
	embeds: z.array(z.never()).nullish(),
```

Reply:
Wouldn't max(0) still be needed

Reply:
Doubt it, you can't pass a value that satisfies `never`, so the predicate should insta reject on the first value passed in.